ts_algorithms/fdk.py

```diff
@@ -0,0 +1,79 @@
+import numpy as np
+import torch
+import tomosipo as ts
+from .fbp import fbp
+
+
+def fdk_weigh_projections(op, projections, recalculate_weights):
+    dev = projections.device
+    u_num_pixels = op.range_shape[2]
+    u_space = (torch.arange(u_num_pixels, dtype=torch.float, device=dev)+0.5-(u_num_pixels/2))
+    v_num_pixels = op.range_shape[0]
+    v_space = (torch.arange(v_num_pixels, dtype=torch.float, device=dev)+0.5-(v_num_pixels/2))
+
+    result = torch.empty(projections.shape, dtype=torch.float, device=dev)
+
+    for i in range(op.range_shape[1]):
+        if recalculate_weights or i == 0:
+            src_to_det = torch.from_numpy(op.range.det_pos[i] - op.range.src_pos[i]).float().to(dev)
+            src_to_obj = torch.from_numpy(-op.range.src_pos[i]).float().to(dev)
+            det_n = torch.from_numpy(op.range.det_normal[i]).float().to(dev)
+            det_u = torch.from_numpy(op.range.det_u[i]).float().to(dev)
+            det_v = torch.from_numpy(op.range.det_v[i]).float().to(dev)
+            det_n_norm = torch.norm(det_n)
+            det_u_norm = torch.norm(det_u)
+            det_v_norm = torch.norm(det_v)
+            src_det_n_dist = torch.dot(src_to_det, det_n) / det_n_norm
+            src_det_u_dist = torch.dot(src_to_det, det_u) / det_u_norm
+            src_det_v_dist = torch.dot(src_to_det, det_v) / det_v_norm
+            src_to_obj_n_dist = torch.dot(src_to_obj, det_n) / det_n_norm
+
+            u_pos_squared = (u_space * det_u_norm + src_det_u_dist)**2
+            v_pos_squared = (v_space * det_v_norm + src_det_v_dist)**2
+
+            weights_mat = src_det_n_dist / torch.sqrt(
+                  u_pos_squared[None, :].expand([v_num_pixels, u_num_pixels])
+                  + v_pos_squared[:, None].expand([v_num_pixels, u_num_pixels])
+                  + src_det_n_dist**2
+            )
+            # scale detector to origin
+            weights_mat *= (src_to_obj_n_dist / src_det_n_dist)
```

Comment:
Scaling invariance does hold up, so that is good! 

``` python 



def test_fdk_scaling_invariance():
    vg = ts.volume(shape=32, size=1)
    pg = ts.cone(
        angles=32, shape=(48, 48), size=(2, 2), src_det_dist=2, src_orig_dist=2
    ).to_vec()

    # Create sinogram and reconstruct at the origin
    A0 = ts.operator(vg, pg)
    x = make_box_phantom()

    rec0 = fdk(A0, A0(x))
    bp0 = A0.T(A0(x))

    s = 2.0
    S = ts.scale(s)
    A = ts.operator(S * vg, S * pg)

    rec = fdk(A, A(x))
    bp = A.T(A(x))

    # Check that backprojection still matches (up to scaling)
    assert torch.allclose(bp / s**2, bp0)
    # Check that fdk still matches (fails!)
    assert torch.allclose(rec, rec0, atol=0.5, rtol=0.01)
```